tools/generators/lib/PBXProj/src/Identifiers.swift

```diff
@@ -49,13 +51,79 @@ public enum Identifiers {
     }
 
     public enum FilesAndGroups {
+        /// The logical type of the element being identified.
+        public enum ElementType: String {
+            /// A normal `PBXFileReference`.
+            case fileReference = "0"
+
+            /// A `PBXGroup`.
+            case group = "1"
+
+            /// A `PBXVariantGroup` or child `PBXFileReference`.
+            case localized = "2"
+
+            /// A `XCVersionGroup` or child `PBXFileReference`.
+            case coreData = "3"
+        }
+
         public static func mainGroup(_ path: String) -> String {
             return #"000000000000000000000003 /* \#(path) */"#
         }
 
         public static let productsGroup = #"""
 000000000000000000000004 /* Products */
 """#
+
+        /// Calculates the identifier for a file or group element at `path`.
+        ///
+        /// - Parameters:
+        ///   - path: The file path for the version group.
+        ///   - type: The type of path being identified.
+        ///   - hashCache: A cache that will be used to guarantee that the
+        ///     identifier returned is unique.
+        public static func element(
+            _ path: String,
+            type: ElementType,
+            hashCache: inout Set<String>
+        ) -> String {
+            let hash = elementHash(path + type.rawValue, hashCache: &hashCache)
+            return #"FF\#(hash) /* \#(path) */"#
+        }
+
+        private static func elementHash(
+            _ hashable: String,
+            hashCache: inout Set<String>
+        ) -> String {
+            var hash: String
+            var retryCount = 0
+            repeat {
+                hash = elementHash(hashable, retryCount: retryCount)
+                retryCount += 1
+            } while hashCache.contains(hash)
+
+            hashCache.insert(hash)
+
+            return hash
+        }
+
+        private static func elementHash(
+            _ path: String,
+            retryCount: Int
+        ) -> String {
+            let content: String
+            if retryCount == 0 {
+                content = path
+            } else {
+                content = "\(path)\0\(retryCount)"
+            }
+
+            let digest = Insecure.MD5.hash(data: content.data(using: .utf8)!)
+            return digest
+                // We can only use the first 22 characters
```

Comment:
IIRC only the first 22 chars are relevant in an MD5 and the rest is padding correct? 

That said, why we have to drop the last 10 characters here for the purpose of generating an identifier later on? Wouldn't it work the same way without dropping even if those characters are not "relevant"?

Reply:
So Xcode identifiers are 24 characters. We are using 2 characters at the front for `"FF"`. That leaves 22 characters that we can use. MD5 generates a digest that is 32 characters long. So we need to truncate it to fit within the remaining 22 characters. We choose the front 22 because are the most unique.

Reply:
I'll add this as a comment.

Reply:
>So Xcode identifiers are 24 characters

Oh ok so this is the bit I was missing, yep a short comment on this would be great 👍